### DISCUSSION TITLE
Creating a 2d ndarray object in bokehjs

### DISCUSSION THREAD

**Discussion by sansseriff** [bokeh/bokeh#13767] [Internals and design (Q&A)]
Hello, my ultimate goal is to create a Bokeh.Image purely in bokehjs. I've learned this needs to be done by constructing a CDS that contains a 2d ndarray object. How do I create the ndarray object? 

So far I've only found a hacky way of creating 1d ndarray objects with minified Bokehjs loaded, using Bokeh.LinAlg.np.arange(0,100,2). 

For anyone interested, this is for some physics calculations [I'm trying out in a svelte app](https://sansseriff.github.io/spdc_cavity_analysis/). I'd like to replace the vanilla canvas part with a bokeh Image. 

Thanks for the help!!

**Comment by bryevdv** [comment]
cc @mattpap it would probably be nice to have a full image example in `bokehjs/examples`

**Comment by sansseriff** [comment]
I'd be happy to make a pull request with a fleshed out example if someone pointed me in the right direction. How to make a 2d ndarray object in JS

**Reply by bryevdv** [reply]
@mattpap can you advise?

**Comment by mattpap** [comment]
ndarrays aren't well defined in bokehjs and the current API is more suited for internal usage. However, the image API requires ndarrays, so here's an example how to define them:

https://github.com/bokeh/bokeh/blob/d9a2f7a036cd7870bce78937f8261f9c8bff72ae/bokehjs/test/integration/regressions.ts#L79-L108

Essentially one has to work with a flat array and do indexing manually, and create an `ndarray` with an appropriate `dtype` and `shape`.

I started PR #13785, where I will add some more examples to bokehjs covering this and other topics.

**Reply by sansseriff** [reply]
The examples in bokehjs/test/integration/examples seem to use the typescript bokeh package, not the minified javascript as served by a CDN. Will this be the recommended way of working with Bokehjs in the future? Downloading as package from npm and bundling it yourself into your app? 

Last time I tried that I had some success, but there were various circular import issues among other things. It might depend on the bundler. With a simple example, I couldn't get bundle size down to something similar to the core minified bundle ("bokeh-3.4.0.min.js").

**Comment by PabloRuizCuevas** [comment]
@sansseriff  You may have been the first person to have integrated bokeh and svelte, I tried for a long time but faced issues everywhere, I can't even import it in sveltekit, could you maybe share some tips on how did you do it?

**Reply by sansseriff** [reply]
If you're asking about using bokeh as a package downloaded from npm, yes its always been problematic. 

I once was able to get something working with and older version of svelte and bokeh. Now that svelte uses vite instead of rollup, it's different. And in the end, it didn't really seem worth it because bundling bokeh directly in my app didn't cut down on the size of the dependency versus using the pre-bundled file (like https://cdn.bokeh.org/bokeh/release/bokeh-3.6.1.min.js). 

I guess the developers only prioritize making bokehjs available as a pre-bundled file. You can import this with a script tag in your main .html file 

I did that here: https://github.com/sansseriff/tag_gui/tree/master/frontend

See in `index.html` the script tags that import the minified files. Then in `src/main.js` how I use `const Bokeh = window.Bokeh;` to make the bokeh namespace available in the rest of the components. 

I've worked mostly with svelte, not sveltekit. But for bokeh to work, you probably need to turn off ssr somewhere, or setup all plots in onMount().

**Comment by bryevdv** [comment]
> I guess the developers only prioritize making bokehjs available as a pre-bundled file. You can import this with a script tag in your main .html file

@sansseriff I can only speak for myself: I just don't have any expertise at all on the JS packaging side, nor any time to gain enough to help with this corner of Bokeh. We'd love to have experienced JS devs join the project and help move this forward. FWIW @ianthomas23 should be embarking on some funded work soon to help improve this story, but truly, the more input from actual JS developers, the better.

**Reply by sansseriff** [reply]
Good to hear. I'm actually on a personal quest to understand and contribute to some other web canvas javascript libraries at the moment. Like konva and motion canvas. I'll keep exploring bokehjs source code. See what I can learn.

**Comment by ianthomas23** [comment]
@sansseriff See #13732 for some of the problems with using BokehJS on its own without Python. I'm planning to fix those problems in the next few months and provide some simple examples of how to use it with various bundlers/frameworks.

**Reply by sansseriff** [reply]
Glad this is in the pipeline. The first problem I encountered with using bokeh from npm was that vite was interpreting the @bokeh/bokeh package as commonjs for some reason, and therefore showing [this type of error](https://discourse.bokeh.org/t/how-to-import-bokehjs/10713). I was able to temporarily fix this by adding `"type": "module",` to the top package.json file in @bokeh/bokehjs. 

After that, I ran in to issues with `connect_signals()` which I believe you are addressing in #13732.
